pkg/instancetype/compatibility/compatibility.go

```diff
@@ -0,0 +1,196 @@
+//nolint:dupl,lll,gocyclo
```

Comment:
Is there any way to satisfy the linters? :)

Reply:
:melting_face: 

I did quickly look but didn't want to get stuck on this tbh, `gocyclo` gives this insane error about `decodeControllerRevisionObject`. Yes it's just one long switch statement but IMHO it's perfectly readable:
```
pkg/instancetype/compatibility/compatibility.go:63:1: cyclomatic complexity 19 of func `decodeControllerRevisionObject` is high (> 15) (gocyclo)
```

`dupl` is also a false positive IMHO, these two functions while logically similar are acting on different types:

```
146-170 lines are duplicate of `pkg/instancetype/compatibility/compatibility.go:172-196` (dupl)
func decodeVirtualMachineInstancetypeSpecRevision(data []byte) *v1alpha1.VirtualMachineInstancetype {
        revision := &v1alpha1.VirtualMachineInstancetypeSpecRevision{}
        strictErr, err := json.UnmarshalStrict(data, revision)
        if err != nil || strictErr != nil {
                // Failed to unmarshal, so the object is not the expected type
                return nil
        }

        instancetypeSpec := &v1alpha1.VirtualMachineInstancetypeSpec{}
        strictErr, err = json.UnmarshalStrict(revision.Spec, instancetypeSpec)
        if err != nil || strictErr != nil {
                return nil
        }

        return &v1alpha1.VirtualMachineInstancetype{
                TypeMeta: metav1.TypeMeta{
                        APIVersion: v1alpha1.SchemeGroupVersion.String(),
                        Kind:       "VirtualMachineInstancetype",
                },
                ObjectMeta: metav1.ObjectMeta{
                        Name: "old-version-of-instancetype-object",
                },
                Spec: *instancetypeSpec,
        }
}
pkg/instancetype/compatibility/compatibility.go:172: 172-196 lines are duplicate of `pkg/instancetype/compatibility/compatibility.go:146-170` (dupl)
func decodeVirtualMachinePreferenceSpecRevision(data []byte) *v1alpha1.VirtualMachinePreference {
        revision := &v1alpha1.VirtualMachinePreferenceSpecRevision{}
        strictErr, err := json.UnmarshalStrict(data, revision)
        if err != nil || strictErr != nil {
                // Failed to unmarshall, so the object is not the expected type
                return nil
        }

        preferenceSpec := &v1alpha1.VirtualMachinePreferenceSpec{}
        strictErr, err = json.UnmarshalStrict(revision.Spec, preferenceSpec)
        if err != nil || strictErr != nil {
                return nil
        }

        return &v1alpha1.VirtualMachinePreference{
                TypeMeta: metav1.TypeMeta{
                        APIVersion: v1alpha1.SchemeGroupVersion.String(),
                        Kind:       "VirtualMachinePreference",
                },
                ObjectMeta: metav1.ObjectMeta{
                        Name: "old-version-of-preference-object",
                },
                Spec: *preferenceSpec,
        }
}
```

`lll` finally is awkward to resolve given the huge auto generated conversion function names, I couldn't find a nice way of breaking these up and tbh I'd rather ignore `lll` in this context and keep them to a single line:

```
pkg/instancetype/compatibility/compatibility.go:70: the line is 163 characters long, which exceeds the maximum of 140 characters. (lll)
        case *v1beta1.VirtualMachineInstancetype, *v1beta1.VirtualMachineClusterInstancetype, *v1beta1.VirtualMachinePreference, *v1beta1.VirtualMachineClusterPreference:
pkg/instancetype/compatibility/compatibility.go:80: the line is 146 characters long, which exceeds the maximum of 140 characters. (lll)
                if err := v1alpha2.Convert_v1alpha2_VirtualMachineClusterInstancetype_To_v1beta1_VirtualMachineClusterInstancetype(obj, dest, nil); err != nil {
pkg/instancetype/compatibility/compatibility.go:92: the line is 142 characters long, which exceeds the maximum of 140 characters. (lll)
                if err := v1alpha2.Convert_v1alpha2_VirtualMachineClusterPreference_To_v1beta1_VirtualMachineClusterPreference(obj, dest, nil); err != nil {
pkg/instancetype/compatibility/compatibility.go:104: the line is 146 characters long, which exceeds the maximum of 140 characters. (lll)
                if err := v1alpha1.Convert_v1alpha1_VirtualMachineClusterInstancetype_To_v1beta1_VirtualMachineClusterInstancetype(obj, dest, nil); err != nil {
pkg/instancetype/compatibility/compatibility.go:116: the line is 142 characters long, which exceeds the maximum of 140 characters. (lll)
                if err := v1alpha1.Convert_v1alpha1_VirtualMachineClusterPreference_To_v1beta1_VirtualMachineClusterPreference(obj, dest, nil); err != nil {
pkg/instancetype/compatibility/compatibility.go:129: the line is 159 characters long, which exceeds the maximum of 140 characters. (lll)
                if err := v1alpha1.Convert_v1alpha1_VirtualMachinePreference_To_v1beta1_VirtualMachinePreference(oldPreferenceObject, newPreferenceObject, nil); err != nil {
pkg/instancetype/compatibility/compatibility.go:137: the line is 167 characters long, which exceeds the maximum of 140 characters. (lll)
                if err := v1alpha1.Convert_v1alpha1_VirtualMachineInstancetype_To_v1beta1_VirtualMachineInstancetype(oldInstancetypeObject, newInstancetypeObject, nil); err != nil {
```

Reply:
ACK